security/advancedtls/crl_test.go

```diff
@@ -0,0 +1,711 @@
+	t.Helper()
+
+	certChain := make([]*x509.Certificate, 0)
+
+	rest, err := ioutil.ReadFile(testdata.Path(name))
+	if err != nil {
+		t.Fatalf("ioutil.ReadFile(%v) failed %v", name, err)
+	}
+	for len(rest) > 0 {
+		var block *pem.Block
+		block, rest = pem.Decode(rest)
+		c, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			t.Fatalf("ParseCertificate error %v", err)
+		}
+		t.Logf("Parsed Cert sub = %v iss = %v", c.Subject, c.Issuer)
+		certChain = append(certChain, c)
+	}
+	return certChain
+}
+
+func loadCRL(t *testing.T, path string) *pkix.CertificateList {
+	b, err := ioutil.ReadFile(testdata.Path(path))
+	if err != nil {
+		t.Fatalf("readFile(%v) failed err = %v", path, err)
+	}
+	crl, err := x509.ParseCRL(b)
+	if err != nil {
+		t.Fatalf("ParseCrl(%v) failed err = %v", path, err)
+	}
+	return crl
+}
+
+func TestCachedCRL(t *testing.T) {
+	cache, err := lru.New(5)
+	if err != nil {
+		t.Fatalf("lru.New: err = %v", err)
+	}
+
+	tests := []struct {
+		desc string
+		val  interface{}
+		ok   bool
+	}{
+		{
+			desc: "Valid",
+			val: &certificateListExt{
+				CertList: &pkix.CertificateList{
+					TBSCertList: pkix.TBSCertificateList{
+						NextUpdate: time.Now().Add(time.Hour),
+					},
+				}},
+			ok: true,
+		},
+		{
+			desc: "Expired",
+			val: &certificateListExt{
+				CertList: &pkix.CertificateList{
+					TBSCertList: pkix.TBSCertificateList{
+						NextUpdate: time.Now().Add(-time.Hour),
+					},
+				}},
+			ok: false,
+		},
+		{
+			desc: "Wrong Type",
+			val:  "string",
+			ok:   false,
+		},
+		{
+			desc: "Empty",
+			val:  nil,
+			ok:   false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			if tt.val != nil {
+				cache.Add(hex.EncodeToString([]byte(tt.desc)), tt.val)
+			}
+			_, ok := cachedCrl([]byte(tt.desc), cache)
+			if tt.ok != ok {
+				t.Errorf("Cache ok error expected %v vs %v", tt.ok, ok)
+			}
+		})
+	}
+}
+
+func TestGetIssuerCRLCache(t *testing.T) {
+	cache, err := lru.New(5)
+	if err != nil {
+		t.Fatalf("lru.New: err = %v", err)
+	}
+
+	tests := []struct {
+		desc      string
+		rawIssuer []byte
+		certs     []*x509.Certificate
+	}{
+		{
+			desc:      "Valid",
+			rawIssuer: makeChain(t, "crl/unrevoked.pem")[1].RawIssuer,
+			certs:     makeChain(t, "crl/unrevoked.pem"),
+		},
+		{
+			desc:      "Unverified",
+			rawIssuer: makeChain(t, "crl/unrevoked.pem")[1].RawIssuer,
+		},
+		{
+			desc:      "Not Found",
+			rawIssuer: []byte("not_found"),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			cache.Purge()
+			_, err := fetchIssuerCRL("test", tt.rawIssuer, tt.certs, RevocationConfig{
+				RootDir: testdata.Path("."),
+				Cache:   cache,
+			})
+			if err == nil && cache.Len() == 0 {
+				t.Error("Verified CRL not added to cache")
+			}
+			if err != nil && cache.Len() != 0 {
+				t.Error("Unverified CRL added to cache")
+			}
+		})
+	}
+}
+
+func TestVerifyCrl(t *testing.T) {
+	tampered := loadCRL(t, "crl/1.crl")
```

Comment:
All test files need to be load using the `testdata` package (e.g. https://github.com/grpc/grpc-go/blob/master/security/advancedtls/advancedtls_integration_test.go#L470), otherwise they will all fail in google3.

Reply:
On L340 loadCRL uses testdata.Path to resolve the argument. If it would help, I can move testdata.Path to all the call sites instead

Reply:
Yeah, Matt, I think that would help a lot. Can you please move `testdata.Path` to every other call sites for now? We can always work on solving any other possible issues if later we find them during the import process. Thanks!